staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/lockingpromise/lockingpromise.go

```diff
@@ -23,57 +23,108 @@ import (
 	cond          sync.Cond
 	activeCounter counter.GoRoutineCounter // counter of active goroutines
-	waitingCount  int                      // number of goroutines idle due to this mutable being unset
+	waitingCount  int                      // number of goroutines idle due to this being unset
 	isSet         bool
 	value         interface{}
 }
 
-var _ promise.LockingMutable = &lockingPromise{}
+func (pr *promisoid) Get() interface{} {
+	pr.lock.Lock()
+	defer pr.lock.Unlock()
+	return pr.GetLocked()
+}
+
+func (pr *promisoid) GetLocked() interface{} {
+	if !pr.isSet {
+		pr.waitingCount++
+		pr.activeCounter.Add(-1)
+		pr.cond.Wait()
+	}
+	return pr.value
+}
+
+func (pr *promisoid) IsSet() bool {
+	pr.lock.Lock()
+	defer pr.lock.Unlock()
+	return pr.IsSetLocked()
+}
+
+func (pr *promisoid) IsSetLocked() bool {
+	return pr.isSet
+}
+
+type writeOnce struct {
+	promisoid
+}
+
+var _ promise.LockingWriteOnce = &writeOnce{}
 
-// NewLockingPromise makes a new promise.LockingMutable
-func NewLockingPromise(lock sync.Locker, activeCounter counter.GoRoutineCounter) promise.LockingMutable {
-	return &lockingPromise{
+// NewWriteOnce makes a new promise.LockingWriteOnce
+func NewWriteOnce(lock sync.Locker, activeCounter counter.GoRoutineCounter) promise.LockingWriteOnce {
+	return &writeOnce{promisoid{
 		lock:          lock,
 		cond:          *sync.NewCond(lock),
 		activeCounter: activeCounter,
+	}}
+}
+
+func (wr *writeOnce) Set(value interface{}) bool {
+	wr.lock.Lock()
+	defer wr.lock.Unlock()
+	return wr.SetLocked(value)
+}
+
+func (wr *writeOnce) SetLocked(value interface{}) bool {
+	if wr.isSet {
+		return false
+	}
+	wr.isSet = true
+	wr.value = value
+	if wr.waitingCount > 0 {
+		wr.activeCounter.Add(wr.waitingCount)
+		wr.waitingCount = 0
+		wr.cond.Broadcast()
 	}
+	return true
 }
 
-func (lp *lockingPromise) Set(value interface{}) {
-	lp.lock.Lock()
-	defer lp.lock.Unlock()
-	lp.SetLocked(value)
+type writeMultiple struct {
+	promisoid
 }
 
-func (lp *lockingPromise) Get() interface{} {
-	lp.lock.Lock()
-	defer lp.lock.Unlock()
-	return lp.GetLocked()
+var _ promise.LockingWriteMultiple = &writeMultiple{}
+
+// NewWriteMultiple makes a new promise.LockingWriteMultiple
+func NewWriteMultiple(lock sync.Locker, activeCounter counter.GoRoutineCounter) promise.LockingWriteMultiple {
+	return &writeMultiple{promisoid{
+		lock:          lock,
+		cond:          *sync.NewCond(lock),
+		activeCounter: activeCounter,
+	}}
 }
 
-func (lp *lockingPromise) SetLocked(value interface{}) {
-	lp.isSet = true
-	lp.value = value
-	if lp.waitingCount > 0 {
-		lp.activeCounter.Add(lp.waitingCount)
-		lp.waitingCount = 0
-		lp.cond.Broadcast()
-	}
+func (wr *writeMultiple) Set(value interface{}) {
+	wr.lock.Lock()
+	defer wr.lock.Unlock()
+	wr.SetLocked(value)
 }
 
-func (lp *lockingPromise) GetLocked() interface{} {
-	if !lp.isSet {
-		lp.waitingCount++
-		lp.activeCounter.Add(-1)
-		lp.cond.Wait()
+func (wr *writeMultiple) SetLocked(value interface{}) {
```

Comment:
Right, the shared stuff can be moved to promisoid.

Reply:
Since I recently changed my mind about which one I need, that could plausibly happen again soon.  Once the client code settles down I might be more interested in trimming.  But you're right, I think it's not a big deal.

Reply:
moved shared SetLocked behavior to promisoid.